sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1085 @@
+	ocspModeFailOpen   = "FAIL_OPEN"
+	ocspModeFailClosed = "FAIL_CLOSED"
+	ocspModeInsecure   = "INSECURE"
+)
+
+const ocspCacheKey = "ocsp_cache"
+
+const (
+	// defaultOCSPResponderTimeout is the total timeout for OCSP responder.
+	defaultOCSPResponderTimeout = 10 * time.Second
+)
+
+const (
+	// cacheExpire specifies cache data expiration time in seconds.
+	cacheExpire = float64(24 * 60 * 60)
+)
+
+type ocspCachedResponse struct {
+	time       float64
+	producedAt float64
+	thisUpdate float64
+	nextUpdate float64
+	status     ocspStatusCode
+}
+
+type Client struct {
+	// caRoot includes the CA certificates.
+	caRoot map[string]*x509.Certificate
+	// certPool includes the CA certificates.
+	certPool              *x509.CertPool
+	ocspResponseCache     *lru.TwoQueueCache
+	ocspResponseCacheLock sync.RWMutex
+	// cacheUpdated is true if the memory cache is updated
+	cacheUpdated bool
+	logFactory   func() hclog.Logger
+}
+
+type ocspStatusCode int
+
+type ocspStatus struct {
+	code ocspStatusCode
+	err  error
+}
+
+const (
+	ocspSuccess                ocspStatusCode = 0
+	ocspStatusGood             ocspStatusCode = -1
+	ocspStatusRevoked          ocspStatusCode = -2
+	ocspStatusUnknown          ocspStatusCode = -3
+	ocspStatusOthers           ocspStatusCode = -4
+	ocspFailedDecomposeRequest ocspStatusCode = -5
+	ocspInvalidValidity        ocspStatusCode = -6
+	ocspMissedCache            ocspStatusCode = -7
+	ocspCacheExpired           ocspStatusCode = -8
+)
+
+// copied from crypto/ocsp.go
+type certID struct {
+	HashAlgorithm pkix.AlgorithmIdentifier
+	NameHash      []byte
+	IssuerKeyHash []byte
+	SerialNumber  *big.Int
+}
+
+// cache key
+type certIDKey struct {
+	HashAlgorithm crypto.Hash
+	NameHash      string
+	IssuerKeyHash string
+	SerialNumber  string
+}
+
+// copied from crypto/ocsp
+var hashOIDs = map[crypto.Hash]asn1.ObjectIdentifier{
+	crypto.SHA1:   asn1.ObjectIdentifier([]int{1, 3, 14, 3, 2, 26}),
+	crypto.SHA256: asn1.ObjectIdentifier([]int{2, 16, 840, 1, 101, 3, 4, 2, 1}),
+	crypto.SHA384: asn1.ObjectIdentifier([]int{2, 16, 840, 1, 101, 3, 4, 2, 2}),
+	crypto.SHA512: asn1.ObjectIdentifier([]int{2, 16, 840, 1, 101, 3, 4, 2, 3}),
+}
+
+// copied from crypto/ocsp
+func getOIDFromHashAlgorithm(target crypto.Hash) (asn1.ObjectIdentifier, error) {
+	for hash, oid := range hashOIDs {
+		if hash == target {
+			return oid, nil
+		}
+	}
+	return nil, fmt.Errorf("no valid OID is found for the hash algorithm: %v", target)
+}
+
+func (c *Client) ClearCache() {
+	c.ocspResponseCache.Purge()
+}
+
+func (c *Client) getHashAlgorithmFromOID(target pkix.AlgorithmIdentifier) crypto.Hash {
+	for hash, oid := range hashOIDs {
+		if oid.Equal(target.Algorithm) {
+			return hash
+		}
+	}
+	// no valid hash algorithm is found for the oid. Falling back to SHA1
+	return crypto.SHA1
+}
+
+// isInValidityRange checks the validity
+func isInValidityRange(currTime, nextUpdate time.Time) bool {
+	return !nextUpdate.IsZero() && !currTime.After(nextUpdate)
```

Comment:
My reading of https://datatracker.ietf.org/doc/html/rfc6960#section-3.2 seems to state that `nextUpdate` is an optional field, i.e., if the OCSP responder doesn't know, they can leave this field blank. 

This matches the ASN.1:

```
      nextUpdate         [0]       EXPLICIT GeneralizedTime OPTIONAL,
```

If that's the case, I'd probably just sub in the default cache expiry period to make things simple, tbh.